code/further_analyses/ica_analysis.py

```diff
@@ -0,0 +1,103 @@
+
+def clean_data():
+  masker = NiftiMasker(smoothing_fwhm=8, memory='nilearn_cache', memory_level=1,
+                       mask_strategy='epi', standardize=False)
+  data_masked_1 = masker.fit_transform(BOLD_file_1)
+  data_masked_2 = masker.fit_transform(BOLD_file_2)
+  data_masked_3 = masker.fit_transform(BOLD_file_3)
+  
+### Apply ICA #################################################################
+from sklearn.decomposition import FastICA
+
+def ica_apply(n_components=20):
+  ica = FastICA(n_components = n_components, random_state = 42)
+  components_masked_1 = ica.fit_transform(data_masked_1).T
+  components_masked_2 = ica.fit_transform(data_masked_2).T
+  components_masked_3 = ica.fit_transform(data_masked_3).T
+
+  # Normalize estimated components, for thresholding to make sense
+  components_masked_1 -= components_masked_1.mean(axis=0)
+  components_masked_1 /= components_masked_1.std(axis=0)
+  components_masked_2 -= components_masked_2.mean(axis=0)
+  components_masked_2 /= components_masked_2.std(axis=0)
+  components_masked_3 -= components_masked_3.mean(axis=0)
+  components_masked_3 /= components_masked_3.std(axis=0)
+  
+  # Threshold
+  components_masked_1[components_masked_1 < .8] = 0
+  components_masked_2[components_masked_2 < .8] = 0
+  components_masked_3[components_masked_3 < .8] = 0
+
+  # Now invert the masking operation, going back to a full 3D representation
+  component_img_1 = masker.inverse_transform(components_masked_1)
+  component_img_2 = masker.inverse_transform(components_masked_2)
+  component_img_3 = masker.inverse_transform(components_masked_3)
+
+### Visualize the results #####################################################
+# Show some interesting components
+from nilearn import image
+from nilearn.plotting import plot_stat_map
+
+def ica_vis(subj_num):
+  # Use the mean as a background
+  mean_img_1 = image.mean_img(BOLD_file_1)
+  mean_img_2 = image.mean_img(BOLD_file_2)
+  mean_img_3 = image.mean_img(BOLD_file_3)
+
+  plot_stat_map(image.index_img(component_img_1, 5), mean_img_1, output_file=os.path.join(data_path,'sub'+subj_num+'_BOLD','task001_run001'+'ica_1'+'.jpg'))
+  plot_stat_map(image.index_img(component_img_1, 12), mean_img_1, output_file=os.path.join(data_path,'sub'+subj_num+'_BOLD','task001_run001'+'ica_2'+'.jpg'))
+
+  plot_stat_map(image.index_img(component_img_2, 5), mean_img_2, output_file=os.path.join(data_path,'sub'+subj_num+'_BOLD','task002_run001'+'ica_1'+'.jpg'))
+  plot_stat_map(image.index_img(component_img_2, 12), mean_img_2, output_file=os.path.join(data_path,'sub'+subj_num+'_BOLD','task002_run001'+'ica_2'+'.jpg'))
+
+plot_stat_map(image.index_img(component_img_3, 5), mean_img_3, output_file=os.path.join(data_path,'sub'+subj_num+'_BO\
+LD','task003_run001'+'ica_1'+'.jpg'))
+  plot_stat_map(image.index_img(component_img_3, 12), mean_img_3, output_file=os.path.join(data_path,'sub'+subj_num+'_BOLD','task003_run001'+'ica_2'+'.jpg'))
+
+
+
+import project_config
```

Comment:
The same comment above applies to this one. I've put these two pca functions into pca_utils.py in the utils folder. So you can directly import and use from that file.